runtime/src/qio/auxFilesys/hdfs/qio_plugin_hdfs.c

```diff
@@ -147,10 +147,45 @@ qioerr hdfs_preadv (void* file, const struct iovec *vector, int count, off_t off
 
   STARTING_SLOW_SYSCALL;
 
+#ifdef HDFS3
+
+  DO_RETAIN(((hdfs_fs*)fs));
+  
+  hdfs_file* hfl = (hdfs_file*)qio_calloc(sizeof(to_hdfs_file(file)), 1);
+  qio_memcpy(hfl, to_hdfs_file(file), sizeof(to_hdfs_file(file)));
+  qio_memcpy(&(hfl->file), &(to_hdfs_file(file)->file), sizeof(hdfsFile));
+  const int pathnmlen = strlen(to_hdfs_file(file)->pathnm);
+  hfl->pathnm = (char*)qio_calloc(sizeof(char), pathnmlen);
+  qio_memcpy(&(hfl->pathnm), &(to_hdfs_File(file)->pathnm), sizeof(char)*pathnmlen);
+
+  hdfs_fs* hfs = (hdfs_fs*)qio_calloc(sizeof(to_hdfs_fs(fs)), 1);
+  qio_memcpy(hfs, to_hdfs_fs(fs), sizeof(to_hdfs_fs(fs)));
+  qio_memcpy(&(hfs->hfs), &(to_hdfs_fs(fs)->hfs), sizeof(hdfsFS));
+  qio_memcpy(&(hfs->fs_name), &(to_hdfs_fs(fs)->fs_name), sizeof(char)*strlen(to_hdfs_fs(fs)->fs_name));
+
+  hfl->file = hdfsOpenFile(hfs->hfs, hfl->pathnm, O_RDONLY, 0, 0, 0);
+
+  //assert connection
+  CREATE_ERROR((hfs->hfs == NULL), err_out, ECONNREFUSED, "Unable to read HDFS file", error);
+
+  if(hfl->file == NULL) {
+    err_out = qio_mkerror_errno();
+    goto error;
+  }
+
+#endif
+
   err_out = 0;
   got_total = 0;
   for(i = 0; i < count; i++) {
+
+#ifndef HDFS3
```

Comment:
Shouldn't this be #ifdef HDFS3?
I thought HDFS3 didn't provide hdfsPread?
How could this possibly work the way it is?